### PR TITLE
embeddings: per repo mutex when getting/downloading an index

### PR DESCRIPTION
**Pull request by novoselrok**
Acquire a mutex for the given repository to serialize access when downloading/getting the repo embedding index. This avoids multiple routines concurrently downloading the same embedding index when the cache is empty, which can lead to an out-of-memory error.

## Test plan

* Tested with a script creating multiple concurrent requests. (Found a deadlock in the original commit that way and fixed it.)

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
looks solid :D nested mutexes scare me a little but if it doesn't lock up in practice, all good.

File: enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go [hunk 1]
```diff
@@ -24,6 +25,38 @@ type repoEmbeddingIndexCacheEntry struct {
 	finishedAt time.Time
 }
 
+// repoMutexMap tracks a mutex for each repository, creating one if it does not yet exist for a given repository ID.
+// This allows concurrent access to the cache for different repositories, while serializing access for a given repository.
+type repoMutexMap struct {
```
Comment (mrnugget) [inline enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go:30]: Take a look at `lockMap` which does a similar thing: https://github.com/sourcegraph/sourcegraph/blob/78123006aec79009cda898c8e4bfb45246da78c1/cmd/github-proxy/shared/shared.go#L223-L252

I think you can get away with an RWMutex here and use a fast and slow path.

Comment (mrnugget) [reply enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go:30]: Here's the diff to update `repoMutexMap` to use the `RWMutex` and API of `lockMap`:

```diff
diff --git a/enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go b/enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go
index 7d11566e7e..a5766d32f1 100644
--- a/enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go
+++ b/enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go
@@ -28,33 +28,31 @@ type repoEmbeddingIndexCacheEntry struct {
 // repoMutexMap tracks a mutex for each repository, creating one if it does not yet exist for a given repository ID.
 // This allows concurrent access to the cache for different repositories, while serializing access for a given repository.
 type repoMutexMap struct {
-	// Mutex prevents concurrent access to the repoMutexes map below.
-	sync.Mutex
+	init        sync.Once
+	mu          sync.RWMutex
 	repoMutexes map[api.RepoID]*sync.Mutex
 }
 
-func (r *repoMutexMap) LockRepo(repoID api.RepoID) {
-	r.Lock()
-	// IMPORTANT: do not use `defer r.Unlock()` because it creates a deadlock.
-	// With defer and 2 concurrent requests, request 1 locks the map, and locks the mutex for repo A.
-	// Request 2 comes in, locks the map, and tries to lock the mutex for repo A, but it has to wait (while still locking the map mutex).
-	// Request 1 finishes processing, tries to unlock the mutex for repo A, but it can't because request 2 locked the access to the map, creating a deadlock.
-	repoMutex, ok := r.repoMutexes[repoID]
-	if !ok {
-		repoMutex = &sync.Mutex{}
-		r.repoMutexes[repoID] = repoMutex
-	}
-	r.Unlock()
+func (m *repoMutexMap) GetLock(repoID api.RepoID) *sync.Mutex {
+	m.init.Do(func() { m.repoMutexes = make(map[api.RepoID]*sync.Mutex) })
 
-	repoMutex.Lock()
-}
+	m.mu.RLock()
+	lock, ok := m.repoMutexes[repoID]
+	m.mu.RUnlock()
 
-func (r *repoMutexMap) UnlockRepo(repoID api.RepoID) {
-	r.Lock()
-	repoMutex := r.repoMutexes[repoID]
-	r.Unlock()
+	if ok {
+		return lock
+	}
+
+	m.mu.Lock()
+	lock, ok = m.repoMutexes[repoID]
+	if !ok {
+		lock = &sync.Mutex{}
+		m.repoMutexes[repoID] = lock
+	}
+	m.mu.Unlock()
 
-	repoMutex.Unlock()
+	return lock
 }
 
 func getCachedRepoEmbeddingIndex(
@@ -67,7 +65,7 @@ func getCachedRepoEmbeddingIndex(
 		return nil, errors.Wrap(err, "creating repo embedding index cache")
 	}
 
-	repoMutexMap := &repoMutexMap{repoMutexes: make(map[api.RepoID]*sync.Mutex)}
+	repoMutexMap := &repoMutexMap{}
 
 	getAndCacheIndex := func(ctx context.Context, repoEmbeddingIndexName embeddings.RepoEmbeddingIndexName, finishedAt *time.Time) (*embeddings.RepoEmbeddingIndex, error) {
 		embeddingIndex, err := downloadRepoEmbeddingIndex(ctx, repoEmbeddingIndexName)
@@ -87,8 +85,9 @@ func getCachedRepoEmbeddingIndex(
 		// Acquire a mutex for the given repository to serialize access.
 		// This avoids multiple routines concurrently downloading the same embedding index
 		// when the cache is empty, which can lead to an out-of-memory error.
-		repoMutexMap.LockRepo(repo.ID)
-		defer repoMutexMap.UnlockRepo(repo.ID)
+		lock := repoMutexMap.GetLock(repo.ID)
+		lock.Lock()
+		defer lock.Unlock()
 
 		lastFinishedRepoEmbeddingJob, err := repoEmbeddingJobsStore.GetLastCompletedRepoEmbeddingJob(ctx, repo.ID)
 		if err != nil {
```

I think that's a bit clearer - the distinction between getting the lock and locking it.

**Review by mrnugget — APPROVED** [review]
Looks good! But nested locks are always a bit hairy, so I personally would copy & paste the `lockMap` implementation we have in github-proxy that has been working for years now 😄 Up to you and your risk tolerance though.

**Comment by mrnugget** [comment]
Hmm, since ultimately this is about concurrent access to the cached values, I wonder whether a `sync.Map` wouldn't be better.

**Comment by novoselrok** [comment]
@mrnugget Thanks for the diff. I decided to take your advice and copy the existing implementation 😄 I added a test to check the implementation, and I verified it failed on `main`.

File: enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go [hunk 2]
```diff
@@ -49,6 +82,13 @@ func getCachedRepoEmbeddingIndex(
 			return nil, err
 		}
 
+		// Acquire a mutex for the given repository to serialize access.
+		// This avoids multiple routines concurrently downloading the same embedding index
+		// when the cache is empty, which can lead to an out-of-memory error.
+		lock := repoMutexMap.GetLock(repo.ID)
```
Comment (camdencheek) [inline enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go:88]: FYI, this sounds like exactly the use case for [singleflight](https://pkg.go.dev/golang.org/x/sync/singleflight)

Comment (novoselrok) [reply enterprise/cmd/embeddings/shared/repo_embedding_index_cache.go:88]: Huh, that looks exactly right 😄

**Pull request merged**